sdk/keyvault/Azure.Security.KeyVault.Shared/src/ChallengeBasedAuthenticationPolicy.cs

```diff
@@ -75,16 +74,18 @@ private async ValueTask ProcessCoreAsync(HttpMessage message, ReadOnlyMemory<Htt
                 message.Request.Content = originalContent;
 
                 // update the cached challenge
-                var challenge = AuthenticationChallenge.GetChallenge(message);
+                challenge = AuthenticationChallenge.GetChallenge(message);
 
-                // if a challenge was returned and its different from the cached _challenge
-                if (challenge != null && !challenge.Equals(_challenge))
+                if (challenge != null)
                 {
-                    // update the cached challenge
-                    _challenge = challenge;
+                    // update the cached challenge if not yet set or different from the current challenge (e.g. moved tenants)
+                    if (_challenge == null || !challenge.Equals(_challenge))
+                    {
+                        _challenge = challenge;
+                    }
 
                     // authenticate the request and resend
-                    await AuthenticateRequestAsync(message, async).ConfigureAwait(false);
+                    await AuthenticateRequestAsync(message, async, challenge).ConfigureAwait(false);
```

Comment:
I think we should only re-authenticate and resend the request when the challenge is different than the cached _challenge, or if _challenge is not set yet. Otherwise we're reauthenticating and resending the request only to get another 401.

Reply:
In other words I think this should be moved into the if block above.

Reply:
I think this will require the partial rewrite we were talking about last night. I can't otherwise lock across async code.

Reply:
we're not locking across this block so moving the async calls into it shouldn't be a problem

Reply:
After chatting offline I think leaving this where it is is the best to start.